Split Chrome and Chromium into separate products

DanielRyanSmith opened the pull request:
Fixes #28970

Changes in this PR:
- Chromium functionality is separated from Chrome and allows the testing of Chromium via `./wpt run chromium ...`
- Chromium can be installed in the virtual environment via `./wpt install chromium {browser|webdriver}`
- Chrome's functionality with `./wpt install chrome browser` has been removed, as it only ever installed Chromium's latest revision.
- Chrome will not attempt to detect a browser binary in the virtual environment, as installation functionality is removed.
- Chrome and Chromium's versions of ChromeDriver are now installed and detected in separate virtual environment `bin` directories `_venv/bin/chrome/chromedriver` and `_venv/bin/chromium/chromedriver` respectively. This is done to avoid either browser detecting and using the wrong webdriver, or needing to reinstall the webdriver when switching between testing Chromium and Chrome.
- When installing ChromeDriver for Chrome, the install command will try to match the version of a browser binary detected or supplied with the `--binary` command regardless of a `--channel` argument being passed.

These are the decision trees that decide the source that Chrome and Chromium will install necessary ChromeDriver versions:
![Chrome_webdriver_decision_tree](https://user-images.githubusercontent.com/56164590/156905890-5b0f0fd0-62a5-412f-92d4-e5d48bbe4323.jpg)

![Chromium_webdriver_decision_tree](https://user-images.githubusercontent.com/56164590/156905894-25ba2f3f-93e7-4e7c-b0d5-4188ac31bb3a.jpg)

DanielRyanSmith committed:
Chromium as a separate product

Chromium will need to be pulled from Chrome's implementation as a separate product. This commit creates the skeleton for Chromium as a separate product, while still being functionally identical to Chrome.

DanielRyanSmith committed:
Separate Chrome and Chromium functionality

Currently, we only install Chromium tip-of-tree using the `./wpt install` command. Because a Chrome binary cannot be installed, we no longer try to find a binary in the virtual environment for Chrome. Addtionally, trying to install Chrome via wpt will raise a NotImplementedError.

Functionality for installing chromedriver is shared, however the chromedriver for each product will now be saved in a folder with their respective product name i.e. `venv/bin/{product}/chromedriver`. This removes the need to detect the wrong chromedriver or reinstall a new chromedriver when switching from testing Chrome to Chromium, or vice-versa.

Some refactoring and moving some actions to private methods has occurred in this commit.

DanielRyanSmith committed:
allow user to supply browser version for install command

the `--browser-version` command line argument is now supported for `./wpt install`. This will allow the user to download a specific version of the Chromium browser binary or chromedriver associated with that version.

DanielRyanSmith committed:
update install tests to reflect split change

Install also now includes a check to make sure that the `--browser-version` install argument can only be supplied with chrome and chromium product.

DanielRyanSmith committed:
increase pipeline timeout

DanielRyanSmith committed:
adjust browser_version install logic

Changing logic for accessing the browser_version command line argument and ensuring that it is not called on browsers that do not support it.

DanielRyanSmith committed:
line and string reformatting

 for Chrome/Chromium classes in browser.py

DanielRyanSmith committed:
ensure browser setup checks venv path for webdrvier

DanielRyanSmith committed:
Refactor run.py to allow browsers to use same setup

Edge, Chrome, and Chromium have very similar browser setup processes. This change refactors the instantiation of browser setup classes to allow multiple browsers to use the same setup class.

DanielRyanSmith committed:
Error handling for checking webdriver version

If a null webdriver path is passed to `webdriver_version`, It will now show a warning and return null.

remove whitespace

small refactoring changes to browser.py

ChromeChromiumBase's private method `_get_chromium_url_by_version` no longer takes in a file name argument to reduce coupling.
- Getting latest chromium revision has been moved to a separate private method, `_get_latest_chromium_revision`.

DanielRyanSmith commented:
@jgraham this PR is rebased onto your changes from [this PR](https://github.com/web-platform-tests/wpt/pull/29089).
I am certain there are changes to be made and better ways to approach some aspects of this. I welcome you and @foolip to scrutinize this PR and chastise me about it where it is deserved 😄

DanielRyanSmith committed:
Add new test for attempting to install Chrome

A new test was added to ensure a NotImplementedError is raised when attempting to install Chrome via `wpt install`

remove line whitespace

remove unused variable

DanielRyanSmith commented on tools/wptrunner/wptrunner/browsers/chromium.py:
```diff
@@ -0,0 +1,150 @@
+from . import chrome_spki_certs
```
This file is almost entirely copied over from `chrome.py` and is very likely in need of some changes.

foolip replied:
Is this still the case?

DanielRyanSmith replied:
Yes it is. I'll go through today and try to discern what might need to be changed or different from the Chrome implementation of this file, but I imagine it's very similar.

foolip commented:
I haven't reviewed the code, but the main thing I think is important to add is the ability to get the Chromium and ChromeDriver binary from the same place, given a revision. For example, the `chrome-linux.zip` and `chromedriver_linux64.zip` here:
https://commondatastorage.googleapis.com/chromium-browser-snapshots/index.html?prefix=Linux_x64/966195/

DanielRyanSmith committed:
Ensure Chromium revision number consistency

This change ensures that if multiple Chromium components are downloaded in a single invocation, all of those components will be downloaded using the same revision number. Affects the browser binary, webdriver, and mojojs.

DanielRyanSmith commented:
> the main thing I think is important to add is the ability to get the Chromium and ChromeDriver binary from the same place, given a revision.

I agree, and I implemented a change to ensure that if multiple components are downloaded from a single invocation, it will be certain that they are pulled from the same revision.

foolip commented on .azure-pipelines.yml:
```diff
@@ -55,6 +55,7 @@ jobs:
   condition: eq(variables['Build.Reason'], 'PullRequest')
   pool:
     vmImage: 'macOS-11'
+  timeoutInMinutes: 180
```
Is this accidentally included in this PR? It's not mentioned in the description.

DanielRyanSmith replied:
Ah yes, this was an action taken when jgraham's PR was not yet merged, so that the CI could run in its entirety to check for errors. I've removed it now.

foolip commented on tools/wpt/browser.py:
```diff
@@ -516,264 +516,201 @@ def version(self, binary=None, webdriver_binary=None):
-                return extracted
-            rmtree(extracted)
+    # The last revision number used to successfully download/install browser binary or webdriver.
+    last_revision_used = None
```
Is this used for Chrome or just Chromium? For Chrome, we can't install the binaries and need to find the driver via https://chromedriver.storage.googleapis.com/LATEST_RELEASE_99 or similar, so I guess just Chromium?

For both this and `revision_attempted`, I wonder if we could keep the information on the file system so that we can check on subsequent runs that we have exactly the binaries that we need?

DanielRyanSmith replied:
I think that sounds reasonable. Perhaps that work can be in the scope of adding the infrastructure to pin the Chromium revision?

foolip commented on tools/wpt/browser.py:
```diff
@@ -516,264 +516,201 @@ def version(self, binary=None, webdriver_binary=None):
-        return "%s%s" % (platform, bits)
+        return f"{self.platform.lower()}{bits}"
 
     def _chromium_platform_string(self):
```
Will this be used by Chrome? If it's just for Chromium, can it be moved into that class?

DanielRyanSmith replied:
Since it is not possible to install a Chrome browser binary via `./wpt install`, then this will be unlikely to be used by Chrome. However, if that were ever to change and we added functionality to install Chrome with wpt, then this would possibly be shared by Chrome and Chromium. Is it just best to isolate all install functionality and attribute it only to Chromium, since as of now we technically _shouldn't_ be installing a chromedriver for Chrome?
I do think that's a good idea, and then if something changes and we supply Chrome installs, there can be some minor refactoring for what is shared. Since supplying Chrome install not on any horizon at the moment, I like this approach.

foolip commented on tools/wpt/browser.py:
```diff
@@ -516,264 +516,201 @@ def version(self, binary=None, webdriver_binary=None):
+        # Make sure we use the same revision in an invocation.
+        # If we have a last revision used, that revision takes priority.
+        if self.last_revision_used is None and revision is None:
+            # No revision provided or previously used. use latest revision.
```
```suggestion
            # No revision provided or previously used. Use the latest revision.
```

foolip commented on tools/wpt/browser.py:
```diff
@@ -516,264 +516,201 @@ def version(self, binary=None, webdriver_binary=None):
-                                   os.path.join(dest, self._chromium_package_name(), "Chromium.app", "Contents", "MacOS"))
-        # find_executable will add .exe on Windows automatically.
-        return find_executable("chrome", os.path.join(dest, self._chromium_package_name()))
+        # The attempted revision is kept. If it is successfully used,
```
It seems like it will be used again even if not successful, is that right? Setting it before it is attempted to be used makes this name a bit funny. Could we have a flow like this?
- Figure out what binaries we want to try to download
- Download those binaries
- Record in the filesystem what exact version those binaries are
- Cache those versions in the Python code if we need to look them up many times during an invocation of `./wpt run`

DanielRyanSmith replied:
I changed some wording to make the logic a bit more readable, but I'll also try to explain. The way it currently works, a revision number will be determined for use and kept in the `revision_attempted` attribute. 
https://github.com/web-platform-tests/wpt/blob/c841a6b1b9dbf77fc1aa5a4ad41a88f148751467/tools/wpt/browser.py#L567-L582

After all other steps are successful (formulating the url, trying to download, unzip, and install the file), then `last_revision_used` is changed to reference the attempted revision number, and that number always takes precedent in any other installs with no revision number passed (browser binary, webdriver, mojojs).
This happens after the browser binary successfully installs:
https://github.com/web-platform-tests/wpt/blob/c841a6b1b9dbf77fc1aa5a4ad41a88f148751467/tools/wpt/browser.py#L711-L716

Or the webdriver successfully installs:
https://github.com/web-platform-tests/wpt/blob/c841a6b1b9dbf77fc1aa5a4ad41a88f148751467/tools/wpt/browser.py#L814-L821

However, If the install attempt had failed, the attempted revision will not be kept in `last_revision_used` for reuse. This is the implementation I came up with to avoid trying to reuse a revision number that had previously failed. I do see that it might read as confusing. I originally only had a `last_revision_used` attribute that was set when the url is created, but it's not unusual to attempt to use a url to download, fail, and then generate another (usually with the latest revision instead).  I wanted to implement a way that the revision is only reused after everything was successful. This is what I came up with, but I'm up for changing it to a better approach.

DanielRyanSmith replied:
After explaining all this, I reimplemented this to not keep revisions attempted and successfully used as attributes, and instead just keep the url that was successfully used for reuse.
https://github.com/web-platform-tests/wpt/blob/615e384f4863ee297d4fc6afde94f02cb5c90b7b/tools/wpt/browser.py#L732-L736
Hopefully this is more intuitive.

foolip commented on tools/wpt/browser.py:
```diff
@@ -516,264 +516,201 @@ def version(self, binary=None, webdriver_binary=None):
-        self.logger.warning("Unable to find the browser binary.")
-        return None
+        # Fall back to the tip-of-tree Chromium build.
+        return f"{self._format_chromium_snapshot_url()}"
```
And this?

```suggestion
        return self._format_chromium_snapshot_url()
```

foolip commented on tools/wpt/browser.py:
```diff
@@ -516,264 +516,201 @@ def version(self, binary=None, webdriver_binary=None):
+        if version and version.lower() != "latest":
+            try:
+                detected_revision = self._get_revision_from_version(version)
+                url = f"{self._format_chromium_snapshot_url(detected_revision)}"
```
This is equivalent, right?
```suggestion
                url = self._format_chromium_snapshot_url(detected_revision)
```

DanielRyanSmith replied:
Yup, this is a remnant from an old implementation adding the file name to the string - cleaned it up!

foolip commented on tools/wpt/browser.py:
```diff
@@ -516,264 +516,201 @@ def version(self, binary=None, webdriver_binary=None):
-    def find_binary(self, venv_path=None, channel=None):
-        if channel == "nightly":
-            return self.find_nightly_binary(self._get_dest(venv_path, channel))
+    def _get_chromium_url_by_version(self, version=None):
```
This is called without a version most of the time. Can you add some documentation explaining what this is for and how it's intended to be used? Because it seems a bit odd, if we have a version number, then why do we need a revision? I think I know the answer in some cases, but it'd be good to write it down.

DanielRyanSmith replied:
That's a good idea. Where is the best place to document this? Just as a comment in the code, or somewhere else? Also, since it is true that you're more likely to _not_ pass the version, this might be better named as something like `_get_chromium_download_url` to feel less version-reliant.

DanielRyanSmith replied:
Additionally, I reimplemented this to not keep revisions attempted and successfully used as attributes, and instead just keep the url that was successfully used for reuse.
https://github.com/web-platform-tests/wpt/blob/615e384f4863ee297d4fc6afde94f02cb5c90b7b/tools/wpt/browser.py#L732-L736
Hopefully this is more intuitive.

DanielRyanSmith committed:
Changes based on review feedback

- f-string changes
- Update logic for revision numbers to be more readable
- Ensure Chrome and Chromium's webdriver installs are handled separately and properly
- Move revision number references to Chromium class
- Move Chromium-specific methods to correct class
- Update tests to use Chromium over Chrome

DanielRyanSmith committed:
Change find_binary handling for Chromium

- Check specific channel directories for Chromium
- Pass filename to _format_chromium_snapshot_url
- Throw request exception when all Chromedriver downloads fail for Chrome

DanielRyanSmith committed:
Chrome installs chromedriver by version arg

- Keep successful download url, not revision
- Raise FileNotFoundError for install_webdriver
- If not browser binary is provided or detected.
= More concise logic for browser setup

foolip commented:
@DanielRyanSmith it would be really helpful with documentation somewhere, ideally in the repo itself, that describes how installing browser, driver and mojojs works for Chromium-based browsers. I'd like to review that and then check if the code seems to match. To take a realistic first-time example, what things are downloaded if running against a Chrome Dev binary installed locally? Is there a fallback if there's no ChromeDriver release yet? And what happens when running a second time, do we quickly check and confirm we have the right binaries, blindly trust that we do, or something else?

DanielRyanSmith committed:
Use _venv3 directory to install/detect binary if it exists (#32835)

DanielRyanSmith committed:
docs describing installation and detection

- Concise commenting and attributes
- Use Chrome's wptrunner functions for Chromium
- Update Chromium BrowserSetup to inherit Chrome's

DanielRyanSmith committed:
Merge branch 'master' into chrome-chromium-split

DanielRyanSmith committed:
Add links to documentation

DanielRyanSmith committed:
make changes to webdriver install decisions

DanielRyanSmith committed:
change docs wording

DanielRyanSmith committed:
update tests

DanielRyanSmith committed:
Move revision detection method into shared class

DanielRyanSmith committed:
explicit typing for BrowserSetup attribute

DanielRyanSmith committed:
correct mypy typing

jgraham reviewed (commented):
I think I need to do another pass over the Chrom[ium] download code, but generally this is looking good :)

DanielRyanSmith committed:
Changes suggest by @jgraham

Small wording changes to comments/docstrings as well.

DanielRyanSmith committed:
add filename to last url used

DanielRyanSmith committed:
Merge branch 'master' into chrome-chromium-split

DanielRyanSmith commented on tools/wpt/browser.py:
```diff
@@ -516,246 +516,179 @@ def version(self, binary=None, webdriver_binary=None):
-            f.write(url)
-        return extracted
+    }.get(uname[0])
+    revision_used = None
```
Yes, it definitely should be - changed!

foolip commented on docs/running-tests/chrome-chromium-installation-detection.md:
```diff
@@ -0,0 +1,100 @@
+
+- **Webdriver**
+: ChromeDriver detection for Chromium will only occur if a valid browser binary has
+been found and the version for that binary is discernible. Once the browser binary version
```
Same question about "discernible" here.

foolip commented on docs/running-tests/chrome-chromium-installation-detection.md:
```diff
@@ -0,0 +1,100 @@
+has been found and the version for that binary is discernible. Once the browser
+binary version is detected, the virtual environment directory will be checked to see
+if a matching ChromeDriver version is already installed. If the browser and ChromeDriver
+versions do not match, the ChromeDriver file will be removed from the directory and
```
```suggestion
versions do not match, the ChromeDriver binary will be removed from the directory and
```

foolip commented on docs/running-tests/chrome-chromium-installation-detection.md:
```diff
@@ -0,0 +1,100 @@
+: ChromeDriver detection for Chromium will only occur if a valid browser binary has
+been found and the version for that binary is discernible. Once the browser binary version
+is detected, the virtual environment directory will be checked to see if a matching ChromeDriver
+version is already installed. If the versions do not match, the ChromeDriver file will be removed
```
```suggestion
version is already installed. If the versions do not match, the ChromeDriver binary will be removed
```

foolip commented on docs/running-tests/chrome-chromium-installation-detection.md:
```diff
@@ -0,0 +1,100 @@
+## Chrome
+
+### Detection
+- **Browser Binary**
```
Let's drop the bullet lists. The bolded text already makes the structure clear, and notes in the middle of the second bullet list don't work.

```suggestion
**Browser**
```

foolip commented on docs/running-tests/chrome-chromium-installation-detection.md:
```diff
@@ -0,0 +1,100 @@
+is no need to reinstall ChromeDriver versions if switching between testing Chrome and Chromium.
+
+### Installation
+- **Browser Binary**
```
```suggestion
**Browser**
```

foolip commented on docs/running-tests/chrome-chromium-installation-detection.md:
```diff
@@ -0,0 +1,100 @@
+Chromium does not have varying channels, so the installation uses the default `nightly`
+designation. The install path is `_venv3/browsers/nightly/{chromium_binary}`.
+
+**Important Note**: If this download process is successful, the Chromium snapshot url
```
For consistency with the previous note, which seems about as important :)

```suggestion
Note: If this download process is successful, the Chromium snapshot URL
```

foolip commented on docs/running-tests/chrome-chromium-installation-detection.md:
```diff
@@ -0,0 +1,100 @@
+
+## Chromium
+
+### Detection
+- **Browser Binary**
+: Chromium browser binary detection is only executed on the virtual environment directory
+`_venv3/browsers/{channel}/`. Chromium’s implementation will not attempt to detect a Chromium
+browser binary on the user’s system outside of this directory. This detection process is only
+used if the user has not passed a binary path as an argument using the `--binary` flag.
```
```suggestion
### Detection

**Browser**: Chromium browser binary detection is only done in the virtual
environment directory `_venv3/browsers/{channel}/`, not on the user’s system
outside of this directory. This detection process is only used if the user has
not passed a binary path as an argument using the `--binary` flag.
```

foolip reviewed (commented):
Love the new documentation on how this all works!

DanielRyanSmith committed:
Update docs/running-tests/chrome-chromium-installation-detection.md

Co-authored-by: Philip Jägenstedt <philip@foolip.org>

DanielRyanSmith committed:
Update docs/running-tests/chrome-chromium-installation-detection.md

Co-authored-by: Philip Jägenstedt <philip@foolip.org>
Update docs/running-tests/chrome-chromium-installation-detection.md

DanielRyanSmith committed:
Apply suggestions from @foolip docs review

Co-authored-by: Philip Jägenstedt <philip@foolip.org>

DanielRyanSmith commented on docs/running-tests/chrome-chromium-installation-detection.md:
```diff
@@ -0,0 +1,100 @@
+
+- **Webdriver**
+: ChromeDriver detection for Chrome will only occur if a valid browser binary
+has been found and the version for that binary is discernible. Once the browser
```
In this context, it meant that we were able to successfully check the version of the binary. Of course, a "valid browser binary" means that we can do this, and so this statement is redundant. I am removing it to avoid confusion.

DanielRyanSmith committed:
Doc formatting and change request

also,
- small syntactic and comment changes
- added TODO comment for Chrome version matching logic

DanielRyanSmith commented on docs/running-tests/chrome-chromium-installation-detection.md:
```diff
@@ -0,0 +1,100 @@
+: ChromeDriver detection for Chrome will only occur if a valid browser binary
+has been found and the version for that binary is discernible. Once the browser
+binary version is detected, the virtual environment directory will be checked to see
+if a matching ChromeDriver version is already installed. If the browser and ChromeDriver
```
Added clarification about version matching to the docs for Chrome and Chromium 👌

DanielRyanSmith commented on docs/running-tests/chrome-chromium-installation-detection.md:
```diff
@@ -0,0 +1,100 @@
+: A version of ChromeDriver will only be installed once a Chrome browser binary
+has been given or detected. A FileNotFoundError will be raised if the user tries
+to download ChromeDriver via the install command and a browser binary is not located.
+`wpt` will install a version of ChromeDriver that version-matches the browser binary.
```
I changed the use of `wpt` in a lot of places here, as I meant to reference the application as a whole. In all of these places, I changed to "WPT" or reworded it entirely.

DanielRyanSmith commented on docs/running-tests/chrome-chromium-installation-detection.md:
```diff
@@ -0,0 +1,100 @@
+through official releases. In this case, the Chromium revision associated with this version is
+detected from the [OmahaProxy API](https://omahaproxy.appspot.com/) and used to download
+Chromium's version of ChromeDriver for use, as this is currently the closest version we can
+match for Chrome Dev. If the version does not match any revision number, the latest revision
```
Yes. I added that to the docs in this paragraph

foolip commented on tools/wpt/browser.py:
```diff
@@ -578,61 +637,235 @@ def install_mojojs(self, dest, channel, browser_binary):
+
+        # Because Chromium and its ChromeDriver should be pulled from the
+        # same revision number, their version numbers should match exactly.
+        if browser_version == chromedriver_version:
```
Exact match for Chromium 🎉

foolip commented on tools/wpt/browser.py:
```diff
@@ -578,61 +637,235 @@ def install_mojojs(self, dest, channel, browser_binary):
+    product = "chrome"
 
+    def _get_webdriver_url(self, version):
+        """Get a ChromeDriver API url to download a version of ChromeDriver that matches
```
```suggestion
        """Get a ChromeDriver API URL to download a version of ChromeDriver that matches
```

foolip commented on docs/running-tests/chrome-chromium-installation-detection.md:
```diff
@@ -0,0 +1,96 @@
+ChromeDriver via `./wpt install` and a browser binary is not located. After browser binary
+detection, a version of ChromeDriver that matches the browser binary will be installed.
+The download source for this ChromeDriver is
+[described here](http://chromedriver.chromium.org/downloads/version-selection).
```
```suggestion
[described here](https://chromedriver.chromium.org/downloads/version-selection).
```

foolip reviewed (approved):
Sorry it took half a century, but looking great!

foolip committed:
Update docs

foolip committed:
Update chrome-chromium-installation-detection.md

DanielRyanSmith commented on tools/wpt/browser.py:
```diff
@@ -514,59 +514,118 @@ def version(self, binary=None, webdriver_binary=None):
+        architecture = self._chromium_platform_string
+
+        # If a specific version is passed as an argument, we will use it.
+        if version is not None and version.lower != "latest":
```
oof, yeah - surprised I missed this and should have caught it in testing. Good catch.

DanielRyanSmith committed:
Apply suggestions from code review

changes suggested by @foolip

Co-authored-by: Philip Jägenstedt <philip@foolip.org>

DanielRyanSmith committed:
Changes based on feedback

- descriptive comment about getting base revision from version
- remove duplicated code

DanielRyanSmith commented on tools/wpt/browser.py:
```diff
@@ -514,59 +514,118 @@ def version(self, binary=None, webdriver_binary=None):
+        # if another component is also installed during this run (browser/webdriver).
+        return f"{url_path}{architecture}/{revision}/"
+
+    def _get_revision_from_version(self, version):
```
I agree - added a comment alongside the name change to explain this.

DanielRyanSmith commented on tools/wpt/browser.py:
```diff
@@ -578,61 +637,235 @@ def install_mojojs(self, dest, channel, browser_binary):
 
-    def _chromedriver_platform_string(self):
-        platform = self.platforms.get(uname[0])
+    def install_webdriver(self, dest=None, channel=None, browser_binary=None):
```
`Browser` has an abstract method  `install_webdriver` and all `Browser` classes have this called externally as a starting point for their webdriver install so I was hesitant to make this change. Maybe there's another way to be more clear here?

DanielRyanSmith committed:
Merge branch 'master' into chrome-chromium-split
